### PR TITLE
Multiple optimizer functionality

### PR DESCRIPTION
**Pull request by anayebi**
- Generalizes the ClipOptimizer to be able to handle multiple optimizer functions:
    - If your loss function returns a list of losses and your learning rate schedule returns a list of learning rates, and you pass in a list of optimizer classes and kwargs of the same length (this is something we check automatically), then it will apply those optimizers to each respectively loss using the respective learning rate schedule.

- Also works in the multi-gpu and minibatched setting

- Prior API is unchanged: you can pass in the usual loss function and optimizer params and it will work. The only minor change is that if you are training with optimizer kwargs (e.g. 'momentum': 0.9), this will no longer be an entry by itself in the optimizer_params, but rather an entry in the field of the optimizer_params known as 'optimizer_kwargs':{'momentum': 0.9}. You will see this slight API change reflected in the unit tests as well.

### CODE REVIEW

**Comment by jvrsgsty** [comment]
When using `train_params['include_global_step'] = True` this crashes. We need to feed the global step into the `optimizer_kwargs` dictionary, instead of simply the `optimizer_params`.

**Comment by anayebi** [comment]
@jvrsgsty I just pushed a change to pass the global_step to the ClipOptimizer and to be able to pass that to the relevant suboptimizers via `include_global_step` in their kwargs.

File: tfutils/optimizer.py [hunk 1]
```diff
@@ -144,54 +215,119 @@ def __init__(self, optimizer, *optimizer_args, **optimizer_kwargs):
...
+            if self.grads_and_vars is None:
+                zero_gvs_list = []
+                for opt_idx, _ in enumerate(minibatch_grads):
+                    zero_gvs_list.insert(opt_idx, self._zero_gvs(opt_idx=opt_idx))
+                self.grads_and_vars = zero_gvs_list
+
+            mini_ops_list = []
+            for opt_idx, curr_mb_gv in enumerate(minibatch_grads):
+                curr_mini_ops = self._mini_ops(grads_and_vars=self.grads_and_vars[opt_idx], minibatch_grads=curr_mb_gv, num_minibatches=num_minibatches)
+                mini_ops_list.extend(curr_mini_ops)
```
Comment (jvrsgsty) [inline tfutils/optimizer.py]: I noticed here `extend` is used to append to the list, whereas in https://github.com/neuroailab/tfutils/pull/146/files#diff-51c7e10eb689dc78b0e9fa24df942647R306 (and elsewhere, when looping over optimizers) `insert` is used. I think this should be consistent throughout the optimizer.

File: tfutils/optimizer.py [hunk 2]
```diff
@@ -213,10 +349,15 @@ def apply_gradients(self, grads_and_vars, global_step=None):
...
+            assert(len(self.grads_and_vars) == len(self._optimizer._optimizer_class))
             with tf.control_dependencies([optimize]):
-                reset_ops = []
-                for grad_v, _ in self.grads_and_vars:
-                    reset_ops.append(tf.assign(grad_v, tf.zeros(grad_v.shape)))
+                if self._multi_mode:
+                    reset_ops = []
+                    for opt_idx, _ in enumerate(self.grads_and_vars):
+                        curr_reset_ops = self._reset_grads(grads_and_vars=self.grads_and_vars[opt_idx])
+                        reset_ops.extend(curr_reset_ops)
```
Comment (jvrsgsty) [inline tfutils/optimizer.py]: Same here, I would change this to `insert` for consistency.
